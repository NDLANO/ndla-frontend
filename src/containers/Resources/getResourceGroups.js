import defined from 'defined';
import {
  RESOURCE_TYPE_LEARNING_PATH,
  RESOURCE_TYPE_SUBJECT_MATERIAL,
  RESOURCE_TYPE_TASKS_AND_ACTIVITIES,
  RESOURCE_TYPE_ASSESSMENT_RESOURCES,
  RESOURCE_TYPE_SOURCE_MATERIAL,
  RESOURCE_TYPE_EXTERNAL_LEARNING_RESOURCES,
} from '../../constants';

export const sortOrder = {
  [RESOURCE_TYPE_LEARNING_PATH]: 1,
  [RESOURCE_TYPE_SUBJECT_MATERIAL]: 2,
  [RESOURCE_TYPE_TASKS_AND_ACTIVITIES]: 3,
  [RESOURCE_TYPE_ASSESSMENT_RESOURCES]: 4,
  [RESOURCE_TYPE_SOURCE_MATERIAL]: 5,
  [RESOURCE_TYPE_EXTERNAL_LEARNING_RESOURCES]: 6,
};

export const groupResourcesByResourceTypes = (
  supplementaryResources,
  coreResources,
) => {
  const resources = [
    ...coreResources,
    ...supplementaryResources
      .map(resource => ({
        ...resource,
        additional: true,
      }))
      .filter(resource => !coreResources.find(core => core.id === resource.id)), // don't show supp resources that exists in core
  ];
  return resources.reduce((obj, resource) => {
    const resourceTypesWithResources = resource.resourceTypes?.map(type => {
      const existing = defined(obj[type.id], []);
      return { ...type, resources: [...existing, resource] };
    });
    const reduced = resourceTypesWithResources?.reduce(
      (acc, type) => ({ ...acc, [type.id]: type.resources }),
      {},
    );
    return { ...obj, ...reduced };
  }, {});
};

export const sortResourceTypes = resourceTypes =>
  [...resourceTypes].sort((a, b) => {
    if (sortOrder[a.id] === undefined && sortOrder[b.id] === undefined)
      return 0;
    if (sortOrder[a.id] === undefined) return 1;
    if (sortOrder[b.id] === undefined) return -1;
    if (sortOrder[a.id] > sortOrder[b.id]) return 1;
    if (sortOrder[a.id] < sortOrder[b.id]) return -1;
    return 0;
  });

export const getResourceGroups = (
  resourceTypes,
  supplementaryResources,
  coreResouces,
) => {
  const groupedResources = groupResourcesByResourceTypes(
    supplementaryResources,
    coreResouces,
  );
  const sortedResourceTypes = sortResourceTypes(resourceTypes);

  return sortedResourceTypes
    .map(type => {
      const resources = defined(groupedResources[type.id], []);
      return { ...type, resources };
    })
    .filter(type => type.resources.length > 0);
};
