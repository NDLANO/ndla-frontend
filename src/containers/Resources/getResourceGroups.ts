/**
 * Copyright (c) 2018-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import {
  RESOURCE_TYPE_LEARNING_PATH,
  RESOURCE_TYPE_SUBJECT_MATERIAL,
  RESOURCE_TYPE_TASKS_AND_ACTIVITIES,
  RESOURCE_TYPE_ASSESSMENT_RESOURCES,
  RESOURCE_TYPE_SOURCE_MATERIAL,
  RESOURCE_TYPE_CONCEPT,
} from "../../constants";
import { GQLResource, GQLResourceType } from "../../graphqlTypes";

export const sortOrder: Record<string, number> = {
  [RESOURCE_TYPE_LEARNING_PATH]: 1,
  [RESOURCE_TYPE_SUBJECT_MATERIAL]: 2,
  [RESOURCE_TYPE_TASKS_AND_ACTIVITIES]: 3,
  [RESOURCE_TYPE_ASSESSMENT_RESOURCES]: 4,
  [RESOURCE_TYPE_SOURCE_MATERIAL]: 5,
  [RESOURCE_TYPE_CONCEPT]: 6,
};

type GQLResourceLike = Pick<GQLResource, "id" | "resourceTypes">;

const groupResourcesByResourceTypes = <T extends GQLResourceLike>(supplementaryResources: T[], coreResources: T[]) => {
  const resources = [
    ...coreResources,
    ...supplementaryResources
      .map((resource) => ({
        ...resource,
        additional: true,
      }))
      .filter((resource) => !coreResources.find((core) => core.id === resource.id)), // don't show supp resources that exists in core
  ];
  return resources.reduce<Record<string, GQLResource[]>>((obj, resource) => {
    const resourceTypesWithResources = resource.resourceTypes?.map((type) => {
      const existing = obj[type.id] ?? [];
      return { ...type, resources: [...existing, resource] };
    });
    const reduced = resourceTypesWithResources?.reduce((acc, type) => ({ ...acc, [type.id]: type.resources }), {});
    return { ...obj, ...reduced };
  }, {});
};

type SharedResourceType = Pick<GQLResourceType, "id" | "name">;

export const sortResourceTypes = (resourceTypes: SharedResourceType[]) =>
  [...resourceTypes].sort((a, b) => {
    if (!sortOrder[a.id] && !sortOrder[b.id]) return 0;
    if (sortOrder[a.id] === undefined) return 1;
    if (sortOrder[b.id] === undefined) return -1;
    if (sortOrder[a.id]! > sortOrder[b.id]!) return 1;
    if (sortOrder[a.id]! < sortOrder[b.id]!) return -1;
    return 0;
  });

export const getResourceGroups = <T extends GQLResourceLike>(
  resourceTypes: SharedResourceType[],
  supplementaryResources: T[],
  coreResouces: T[],
): GQLResourceType[] => {
  const groupedResources = groupResourcesByResourceTypes(supplementaryResources, coreResouces);
  const sortedResourceTypes = sortResourceTypes(resourceTypes);

  return sortedResourceTypes
    .map((type) => {
      const resources = groupedResources[type.id] ?? [];
      return { ...type, resources };
    })
    .filter((type) => type.resources.length > 0);
};
