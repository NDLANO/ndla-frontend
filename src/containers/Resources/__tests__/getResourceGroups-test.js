/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { resourceData1, resourceData2, resourceTypes } from "./mockResources";
import { getResourceGroups, sortResourceTypes } from "../getResourceGroups";

test("get core resources grouped by types", () => {
  const groups = getResourceGroups(resourceTypes, [], resourceData1);

  expect(groups).toMatchSnapshot();
});

test("get core and supplementary resources grouped by types", () => {
  const groups = getResourceGroups(resourceTypes, resourceData2, resourceData1);

  expect(groups).toMatchSnapshot();
});

test("resources types sort order", () => {
  const types = sortResourceTypes(resourceTypes);

  expect(types.length).toBe(6);
  expect(types[0].id).toBe("urn:resourcetype:learningPath");
  expect(types[1].id).toBe("urn:resourcetype:subjectMaterial");
  expect(types[2].id).toBe("urn:resourcetype:tasksAndActivities");
  expect(types[3].id).toBe("urn:resourcetype:reviewResource");
  expect(types[4].id).toBe("urn:resourcetype:SourceMaterial");
  expect(types[6].id).toBe("urn:resourcetype:concept");

  expect(types).toMatchSnapshot();
});

test("filter out duplicates", () => {
  const dupe = resourceData2[0];
  const groups = getResourceGroups(resourceTypes, resourceData2, [...resourceData1, dupe]);
  const type = groups.find((group) => group.id === "urn:resourcetype:subjectMaterial");
  expect(type.resources.reduce((acc, resource) => (resource.id === dupe.id ? [...acc, dupe] : acc), []).length).toBe(1);
});
