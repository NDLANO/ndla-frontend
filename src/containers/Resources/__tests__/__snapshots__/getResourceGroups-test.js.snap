// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get core and supplementary resources grouped by types 1`] = `
[
  {
    "id": "urn:resourcetype:learningPath",
    "name": "Læringssti",
    "resources": [
      {
        "contentUri": "urn:article:1",
        "id": "urn:resource:2",
        "name": "Hva er en idé?",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
          {
            "id": "urn:resourcetype:learningPath",
          },
        ],
      },
    ],
  },
  {
    "id": "urn:resourcetype:subjectMaterial",
    "name": "Fagstoff",
    "resources": [
      {
        "contentUri": "urn:learningpath:1",
        "id": "urn:resource:1",
        "name": "Teknikker for idéutvikling",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
        ],
      },
      {
        "contentUri": "urn:article:1",
        "id": "urn:resource:2",
        "name": "Hva er en idé?",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
          {
            "id": "urn:resourcetype:learningPath",
          },
        ],
      },
      {
        "contentUri": "urn:article:2",
        "id": "urn:resource:3",
        "name": "Ideer og idéutvikling",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
          {
            "id": "urn:resourcetype:academicArticle",
          },
        ],
      },
      {
        "additional": true,
        "contentUri": "urn:article:3",
        "id": "urn:resource:4",
        "name": "Hva er menneskerettigheter?",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
        ],
      },
      {
        "additional": true,
        "contentUri": "urn:article:2",
        "id": "urn:resource:6",
        "name": "Endring av NATOs strategi",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
          {
            "id": "urn:resourcetype:academicArticle",
          },
        ],
      },
    ],
    "subtypes": [
      {
        "id": "urn:resourcetype:academicArticle",
        "name": "Fagartikkel",
      },
      {
        "id": "urn:resourcetype:movieAndClip",
        "name": "Film og filmklipp",
      },
      {
        "id": "urn:resourcetype:simulation",
        "name": "Simulering",
      },
    ],
  },
  {
    "id": "urn:resourcetype:tasksAndActivities",
    "name": "Oppgaver og aktiviteter",
    "resources": [
      {
        "additional": true,
        "contentUri": "urn:article:4",
        "id": "urn:resource:5",
        "name": "Påstander konflikter",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:tasksAndActivities",
          },
        ],
      },
    ],
    "subtypes": [
      {
        "id": "urn:resourcetype:exercise",
        "name": "Øvelse",
      },
      {
        "id": "urn:resourcetype:task",
        "name": "Oppgave",
      },
      {
        "id": "urn:resourcetype:workAssignment",
        "name": "Arbeidsoppdrag",
      },
    ],
  },
]
`;

exports[`get core resources grouped by types 1`] = `
[
  {
    "id": "urn:resourcetype:learningPath",
    "name": "Læringssti",
    "resources": [
      {
        "contentUri": "urn:article:1",
        "id": "urn:resource:2",
        "name": "Hva er en idé?",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
          {
            "id": "urn:resourcetype:learningPath",
          },
        ],
      },
    ],
  },
  {
    "id": "urn:resourcetype:subjectMaterial",
    "name": "Fagstoff",
    "resources": [
      {
        "contentUri": "urn:learningpath:1",
        "id": "urn:resource:1",
        "name": "Teknikker for idéutvikling",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
        ],
      },
      {
        "contentUri": "urn:article:1",
        "id": "urn:resource:2",
        "name": "Hva er en idé?",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
          {
            "id": "urn:resourcetype:learningPath",
          },
        ],
      },
      {
        "contentUri": "urn:article:2",
        "id": "urn:resource:3",
        "name": "Ideer og idéutvikling",
        "resourceTypes": [
          {
            "id": "urn:resourcetype:subjectMaterial",
          },
          {
            "id": "urn:resourcetype:academicArticle",
          },
        ],
      },
    ],
    "subtypes": [
      {
        "id": "urn:resourcetype:academicArticle",
        "name": "Fagartikkel",
      },
      {
        "id": "urn:resourcetype:movieAndClip",
        "name": "Film og filmklipp",
      },
      {
        "id": "urn:resourcetype:simulation",
        "name": "Simulering",
      },
    ],
  },
]
`;

exports[`resources types sort order 1`] = `
[
  {
    "id": "urn:resourcetype:learningPath",
    "name": "Læringssti",
  },
  {
    "id": "urn:resourcetype:subjectMaterial",
    "name": "Fagstoff",
    "subtypes": [
      {
        "id": "urn:resourcetype:academicArticle",
        "name": "Fagartikkel",
      },
      {
        "id": "urn:resourcetype:movieAndClip",
        "name": "Film og filmklipp",
      },
      {
        "id": "urn:resourcetype:simulation",
        "name": "Simulering",
      },
    ],
  },
  {
    "id": "urn:resourcetype:tasksAndActivities",
    "name": "Oppgaver og aktiviteter",
    "subtypes": [
      {
        "id": "urn:resourcetype:exercise",
        "name": "Øvelse",
      },
      {
        "id": "urn:resourcetype:task",
        "name": "Oppgave",
      },
      {
        "id": "urn:resourcetype:workAssignment",
        "name": "Arbeidsoppdrag",
      },
    ],
  },
  {
    "id": "urn:resourcetype:reviewResource",
    "name": "Vurderingsressurs",
  },
  {
    "id": "urn:resourcetype:SourceMaterial",
    "name": "Kildemateriale",
  },
  {
    "id": "urn:resourcetype:concept",
    "name": "Forklaring",
  },
]
`;
